cargo-miri/src/phases.rs

```diff
@@ -105,149 +105,148 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
-    // In nextest we have to also forward the main `verb`.
-    if cargo_cmd == "nextest" {
-        cmd.arg(
-            args.next()
-                .unwrap_or_else(|| show_error!("`cargo miri nextest` expects a verb (e.g. `run`)")),
-        );
-    }
-    // We set the following flags *before* forwarding more arguments.
-    // This is needed to fix <https://github.com/rust-lang/miri/issues/2829>: cargo will stop
-    // interpreting things as flags when it sees the first positional argument.
-
-    // Make sure the build target is explicitly set.
-    // This is needed to make the `target.runner` settings do something,
-    // and it later helps us detect which crates are proc-macro/build-script
-    // (host crates) and which crates are needed for the program itself.
-    if get_arg_flag_value("--target").is_none() {
-        // No target given. Explicitly pick the host.
+        // We always setup.
+        let miri_sysroot = setup(&subcommand, target.as_str(), &rustc_version, verbose, quiet);
+
+        // Invoke actual cargo for the job, but with different flags.
+        // We re-use `cargo test` and `cargo run`, which makes target and binary handling very easy but
+        // requires some extra work to make the build check-only (see all the `--emit` hacks below).
+        // <https://github.com/rust-lang/miri/pull/1540#issuecomment-693553191> describes an alternative
+        // approach that uses `cargo check`, making that part easier but target and binary handling
+        // harder.
+        let cargo_miri_path = env::current_exe()
+            .expect("current executable path invalid")
+            .into_os_string()
+            .into_string()
+            .expect("current executable path is not valid UTF-8");
+        let cargo_cmd = match subcommand {
+            MiriCommand::Forward(s) => s,
+            MiriCommand::Setup => return, // `cargo miri setup` stops here.
+            MiriCommand::Clean => unreachable!(),
+        };
+        let metadata = get_cargo_metadata();
+        let mut cmd = cargo();
+        cmd.arg(&cargo_cmd);
+        // In nextest we have to also forward the main `verb`.
+        if cargo_cmd == "nextest" {
+            cmd.arg(args.next().unwrap_or_else(|| {
+                show_error!("`cargo miri nextest` expects a verb (e.g. `run`)")
+            }));
+        }
+        // We set the following flags *before* forwarding more arguments.
+        // This is needed to fix <https://github.com/rust-lang/miri/issues/2829>: cargo will stop
+        // interpreting things as flags when it sees the first positional argument.
+
+        // Make sure the build target is explicitly set.
+        // This is needed to make the `target.runner` settings do something,
+        // and it later helps us detect which crates are proc-macro/build-script
+        // (host crates) and which crates are needed for the program itself.
         cmd.arg("--target");
-        cmd.arg(host);
-    }
+        cmd.arg(target);
```

Comment:
This does not seem right -- won't this pass `--target` twice if there's also a target set in `args`?

CI passes though so maybe cargo just ignores duplicate targets?

Reply:
I changed the code to only pass `--target` if it was not specified at all.